ProjFS.Mac/PrjFSKext/PrjFSKext/PrjFSProviderUserClient.cpp

```diff
@@ -95,16 +105,21 @@ void PrjFSProviderUserClient::free()
 // Called when the user process explicitly or implicitly (process death) closes
 // the connection.
 IOReturn PrjFSProviderUserClient::clientClose()
+{
+    this->cleanupProviderRegistration();
```

Comment:
Will `PrjFSProviderUserClient::stop` always be called?  

(If it is I'm wondering if we need to call `cleanupProviderRegistration()`)

Reply:
Similarly to the `free` question - yes, we *could* wait for `stop()` but `clientClose()` will definitely be called in the context of the disconnecting or dying client, whereas I don't think there's any such guarantee about `stop()` on user clients. (Indeed, it wouldn't be called at all if we didn't `terminate(0)`.) This means that potentially, a client that disconnects and immediately reconnects to the same virtualisation root could find that its previous provider client connection hasn't been deregistered yet because `stop()` hasn't been called on it yet. The chances of this happening are minimal, but I'd rather avoid the situation. Plus, we'd have to guarantee that `sendMessage()` will be safe to call past `clientClose()`, which is just an unnecessary extra burden.